### PR TITLE
Post inject article added

### PR DESCRIPTION
**Pull request by TomasVotruba**

### CODE REVIEW

File: source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+use Doctrine\ORM\EntityManagerInterface;
+
+abstract class AbstractBaseRepository
+{
+    /**
+     * @var EntityManagerInterface
+     */
+    protected $entityManager;
+    
+    public function setEntityManager(EntityManagerInterface $entityManager)
```
Comment (fprochazka) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: I was hoping you'd point out why is this bad, now show more examples how to use it more easily.

File: source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+
+namespace App\Presenter;
+
+abstract class AbstractBasePresenter extends Nette\Application\UI\Presenter
+{
+    /**
+     * @inject
+     * @var Translator
+     */
+    public $translator;
```
Comment (fprochazka) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: this should also contain a disclaimer, that it's pragmatic over clean design

image

File: source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+}
+```
+
+And I can use it `ProductPresenter` alongside with constructor injection
```
Comment (enumag) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: use it in ... along with constructor injection

File: source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+
+services:
+    -
+        class: App\Repository\ProductRepository
+            inject: on
+```
+
+## Your Code is Seen as Manual How to Write 
+
+Why? Because "what you see is what you write". New programmer joins the teams, sees this handy @inject feature and uses when possible and handy.
```
Comment (enumag) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: ` around inject?

File: source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+    private $productRepository;
+    
+    public function __construct(ProductRepository $productRepository) 
+    {
+        $this->productRepository = $productRepository; 
+    }
+}
+```
+
+This is quite clean and easy to use, because presenters in Nette are [injected to by default](https://github.com/nette/application/blob/3165d3a8dab876f4364cdcba450a33ab0182049a/src/Bridges/ApplicationDI/ApplicationExtension.php#L111-L116).
```
Comment (enumag) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: "are injected to" => "have `@inject` annotations enabled"

also the code above uses ctor injection, not inject annotations - you're mixing two different things here

File: source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+            - setEntityManager
+    - 
+        class: App\Repository\CategoryRepository
+        setup:
+            - setEntityManager
+```           
+
+### SO much writing! 
+
+It is clean but with so much writing? Thanks, but no, thanks. Let's go back to `@inject`...
```
Comment (enumag) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: No it's not exactly clean. Setter is a bit better than public property but still not exactly clean because the object still might not be fully initialized.

Also in my opinion inject* methods are bit better than set* methods because setter suggests that the service is not stateless (not the case here) where as inject* by convention are methods that should be called only once and immediately after constructor.

(Of course with inject* methods you don't need the decorator to add "setup" but to add the inject tag.)

File: source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+    # or do you need to call "setTranslator" on every component?
+    App\Component\AbstractBaseControl:
+        setup:
+            - setTranslator
+```
+
+That's it!
+
+
+## What have you learn today?
```
Comment (enumag) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: have you learned / did you learn - I'm not exactly sure which

File: source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+
+- It easy to overuse `@inject` in places where it doesn't solve any problem
```
Comment (enumag) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: It is easy

File: source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+- The problem `@inject`/`inject<method>` method were born to solve is called *dependency hell*
+- If you need to **decorate service of some type**, use *Decorator Extension*
+- This will lead **to better framework understandability and usability**
+- ...and world peace in time :)
+
+In next article, we will look on other practical use cases for Decorator Extension.
```
Comment (enumag) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: look at

File: source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+// app/Repository/ProductRepository.php
+
+namespace App\Repository;
+
+use App\Model\Product\ProductSorer;
+
+final ProductRepository extends AbstractBaseRepository
+{
+    /**
+     * @var ProductSorer
```
Comment (enumag) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: sorter

File: source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+```
+
+And specific repository with some dependency:
+
+```language-php
+// app/Repository/ProductRepository.php
+
+namespace App\Repository;
+
+use App\Model\Product\ProductSorer;
```
Comment (enumag) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: sorter

File: source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md [hunk 12]
```diff
@@ -0,0 +1,254 @@
+---
+layout: post
+title: "How to avoid @inject thanks to Decorator feature in Nette"
+perex: "I often find <code>@inject</code> being overused in projects I review while mentoring. They often bring less writing, but in exchange they break SOLID principals. Today I will show you solution that will <strong>keep your code both small and clean</strong> - <strong>Decorator feature in Nette</strong>."
```
Comment (enumag) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: SOLID should be link to wiki

File: source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+
+As [Derek Simons says](https://www.ted.com/talks/simon_sinek_how_great_leaders_inspire_action) says...
+
+## ...Start with "Why"
+
+Why am I writing this article? I try to improve knowledge interoperability between frameworks by finding the common lowest denominator, so it **is easier to understand and use each other**.
+
+I understand every framework has its own specific way how to solve same issues, but it creates *knowledge vendor lock*. So even if I know how to solve some problem in Nette, I cannot do that in Symfony and have to re-learn it again.
+ 
+Today, I will try to agree on setter injection with you.
```
Comment (enumag) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: This entire section left me more confused than anything else. Not sure what you intended to say.

Comment (TomasVotruba) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: This should be probably standalone article, hmm (todo).

Let me rephrase:

- magic injection is handled in every framework by different way (@inject, inject<>, Kdyby\Autowired in Nette x service locator, JMSDiExtra bundle in Symfony), different addons, bundles or extensions
- **yet they have the same goal**: get some dependency without using constructor injection
- I want to suggest to take common way, that is usable both in Symfony and Nette

Better?

File: source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md (hunk 4 above)
Comment (TomasVotruba) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: Yep, thanks

File: source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md (hunk 3 above)
Comment (TomasVotruba) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: Yep

File: source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md (hunk 2 above)
Comment (TomasVotruba) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: What does it mean and what should that imply?

File: source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md (hunk 5 above)
Comment (TomasVotruba) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: They also have `inject<>` methods enabled.

Where should this be placed then? It is related to `AbstractBasePresenter` class

File: source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md (hunk 1 above)
Comment (TomasVotruba) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: Suggestion?

File: source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md (hunk 6 above)
Comment (TomasVotruba) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: As I stated in the beginning, I try to find common language between frameworks.
Nette uses `inject<>`, but everybody else uses setter injection via `set<>`.

If I would relate to Nette context only, I would agree.

Comment (TomasVotruba) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: How do you suggest rephrase this?

File: source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md (hunk 7 above)
Comment (TomasVotruba) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: Yes, thank you. "have you learne*d" - in meaning of learned and still keeping the knowledge.

File: source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md (hunk 12 above)
Comment (enumag) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: No I don't. Never really needed it.

File: source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md (hunk 13 above)
Comment (enumag) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: Meaning set* methods + some service setup decorator implementation? This article actually feels a lot nette-specific since it mentions inject annotations several times. Or better said moving from nette-specific things to something that may be common for multiple frameworks - that lowest denominator you mentioned is apparently set* methods solution in this case? If you rewrite it to say that the goal is to discourage nette- (or framework-) specific things in favor of those that may be common it might be better.

File: source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md (hunk 5 above)
Comment (enumag) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: Oh, right. I kind of lost context as I was going back and forth in the article. Nevermind, it's ok when reading it normally. Just rephrase the "are injected to" thing, that's a bit unclear.

File: source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md (hunk 6 above)
Comment (enumag) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: After thinking about this for a bit - just change "clean" to "cleaner" and it should be ok. No need to go into the details I mentioned here.

File: source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md (hunk 1 above)
Comment (TomasVotruba) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: + This usually takes lot of explaining that needs broader view - longtime project, cooperation with other packages, programmers etc. Depends on one's situation.

I'll try to explain in comments, if required.

File: source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+
+## Your Code is Seen as Manual How to Write 
+
+Why? Because "what you see is what you write". New programmer joins the teams, sees this handy `@inject` feature and uses when possible and handy.
+
+Some of you, who already talked about `@inject` method usage already there are some and only few specific places to use it.
+
+## Where to only @inject?
+
+**To prevent constructor hell**. If you meet this term first time, go read [this short explanation](https://phpfashion.com/di-a-predavani-zavislosti#toc-constructor-hell) from David Grudl.
```
Comment (Achse) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: I would suggest to point out here: Do not extend and you'll never face this problem, ever..

Comment (Achse) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: Of course, frameworks have controllers designed to be extened and you have no choice in this case, but I would make a complaint about taht design here.

File: source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md (hunk 13 above)
Comment (TomasVotruba) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: Well said. You should be writing too!

File: source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md (hunk 14 above)
Comment (TomasVotruba) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: This is good point, but stretching to much from current level of know-how.

I would love to read an article about app with lots of Presenters in Nette and without `AbstractBasePresenter` approach.

I'm ready to review, if you decide to write one :)

Comment (TomasVotruba) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: Feel free to create issue in nette/application. It's too big goal for me here.

**Comment by TomasVotruba** [comment]
Merged, but **still accepting comments**!

File: source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md (hunk 14 above)
Comment (TomasVotruba) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: I actually try not to complain, since I see no added value in it apart relieving emotions.

Rather I try to suggest realistic path to better design, that might lead to solution to your note in the future.

Comment (Achse) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: I agree this is maybe more for new article. And I feel it not as an emotion, but sharing information what is really root of this problem.

Comment (TomasVotruba) [inline source/_posts/2016/2016-12-24-how-to-avoid-inject-thanks-to-decorator-feature-in-nette.md]: Just a note for others: We continued in private conversation.

Feel free to complete any info.
